### PR TITLE
🥅   add a validation on the `access_token` signature on refresh

### PR DESCRIPTION
**Pull request by shkurata**

### CODE REVIEW

File: api/src/api/openid.ts [hunk 1]
```diff
@@ -136,11 +142,20 @@ export const openIdRequestHook: onRequestAsyncHookHandler = async (request) => {
...
+        await jwtVerify(request.session.openId.tokenSet.access_token, JWKS);
+      }
     } catch (error) {
       if (
-        !(error instanceof Error) ||
-        !error.message.startsWith('invalid_grant')
+        (!(error instanceof Error) ||
+          !error.message.startsWith('invalid_grant')) &&
+        !(error instanceof errors.JOSEError)
       ) {
```
Comment (shkurata) [inline api/src/api/openid.ts:159]: Maybe this condition can be better written. Any ideas?

File: api/src/api/openid.ts [hunk 2]
```diff
@@ -136,11 +142,20 @@ export const openIdRequestHook: onRequestAsyncHookHandler = async (request) => {
...
-        await request.server.openId.client.refresh(new TokenSet(tokenSet));
+        await request.server.openId.client.refresh(tokenSet);
+      if (request.session.openId.tokenSet.access_token != null) {
+        const JWKS = createRemoteJWKSet(
+          new URL(
+            `${request.server.config.AUTH_ISSUER}/protocol/openid-connect/certs`,
+          ),
+        );
+        await jwtVerify(request.session.openId.tokenSet.access_token, JWKS);
+      }
```
Comment (shkurata) [inline api/src/api/openid.ts:153]: The `jwtVerify` method throws an error on invalid signature and that should be enough. We still have the option to add more checks.

**Comment by shkurata** [comment]
How can this be tested?

File: api/src/api/openid.ts [hunk 3]
```diff
@@ -136,11 +142,20 @@ export const openIdRequestHook: onRequestAsyncHookHandler = async (request) => {
   if (tokenSet.expired()) {
     try {
       request.session.openId.tokenSet =
-        await request.server.openId.client.refresh(new TokenSet(tokenSet));
+        await request.server.openId.client.refresh(tokenSet);
```
Comment (Moeriki) [inline api/src/api/openid.ts:145]: We'd probably want to keep the fetched tokenSet in a separate variable until verified. As it is now we store the unvalidated token in the session.

**Review by Moeriki — CHANGES_REQUESTED** [review]

**Comment by Moeriki** [comment]
Validation needs to happen on the [connect callback](https://github.com/OpenSkool/OpenSkool/blob/main/api/src/api/openid.ts#L84) as well.

File: api/src/api/openid.ts [hunk 4]
```diff
@@ -136,11 +142,20 @@ export const openIdRequestHook: onRequestAsyncHookHandler = async (request) => {
...
+      if (request.session.openId.tokenSet.access_token != null) {
```
[the remark this answers is not in the document]
Comment (shkurata) [reply api/src/api/openid.ts:146]: Do you mean that we should make an interface for it?

File: api/src/api/openid.ts [hunk 5]
```diff
@@ -136,11 +142,20 @@ export const openIdRequestHook: onRequestAsyncHookHandler = async (request) => {
...
+        const JWKS = createRemoteJWKSet(
```
[the remark this answers is not in the document]
Comment (shkurata) [reply api/src/api/openid.ts:147]: I have the `app` in order to do that. Any ideas?

**Comment by Moeriki** [comment]
> How can this be tested?

Manually 😄 the complexity of an OAuth flow is beyond the sensibilities of unit testing and cannot be integration tested.

In good time when we add E2E tests this will be covered well enough.

**Review by Moeriki — APPROVED** [review]

**Comment by shkurata** [comment]
> Validation needs to happen on the [connect callback](https://github.com/OpenSkool/OpenSkool/blob/main/api/src/api/openid.ts#L84) as well. 

We have there the whole `tokenSet`. Should we do it for every JWT in the set?

**Comment by Moeriki** [comment]
Manually merged locally.

**Pull request closed without merging**